src/presentation/service/services.go

```diff
@@ -0,0 +1,470 @@
+	}
+
+	var versionPtr *valueObject.ServiceVersion
+	if input["version"] != nil {
+		log.Printf("Version: %v", input["version"])
+		version, err := valueObject.NewServiceVersion(input["version"])
+		if err != nil {
+			return NewServiceOutput(UserError, err.Error())
+		}
+		versionPtr = &version
+	}
+
+	portBindings := []valueObject.PortBinding{}
+	if _, exists := input["portBindings"]; exists {
+		rawPortBindings, assertOk := input["portBindings"].([]string)
+		if !assertOk {
+			return NewServiceOutput(UserError, "InvalidPortBindings")
+		}
+
+		for _, rawPortBinding := range rawPortBindings {
+			portBinding, err := valueObject.NewPortBinding(rawPortBinding)
+			if err != nil {
+				continue
+			}
+			portBindings = append(portBindings, portBinding)
+		}
+	}
+
+	autoCreateMapping := true
+	if input["autoCreateMapping"] != nil {
+		var err error
+		autoCreateMapping, err = voHelper.InterfaceToBool(
+			input["autoCreateMapping"],
+		)
+		if err != nil {
+			return NewServiceOutput(UserError, "InvalidAutoCreateMapping")
+		}
+	}
+
+	dto := dto.NewCreateCustomService(
+		name, svcType, startCmd, []valueObject.ServiceEnv{}, portBindings,
+		nil, nil, nil, nil, nil, versionPtr, nil, nil, nil, nil, nil, nil, nil, nil,
+		&autoCreateMapping,
+	)
+
+	servicesQueryRepo := servicesInfra.NewServicesQueryRepo(service.persistentDbService)
+	servicesCmdRepo := servicesInfra.NewServicesCmdRepo(service.persistentDbService)
+	mappingQueryRepo := mappingInfra.NewMappingQueryRepo(service.persistentDbService)
+	mappingCmdRepo := mappingInfra.NewMappingCmdRepo(service.persistentDbService)
+	vhostQueryRepo := vhostInfra.NewVirtualHostQueryRepo(service.persistentDbService)
+
+	err = useCase.CreateCustomService(
+		servicesQueryRepo, servicesCmdRepo, mappingQueryRepo,
+		mappingCmdRepo, vhostQueryRepo, dto,
+	)
+	if err != nil {
+		return NewServiceOutput(InfraError, err.Error())
+	}
+
+	return NewServiceOutput(Success, "CustomServiceCreated")
+}
+
+func (service *ServicesService) Update(input map[string]interface{}) ServiceOutput {
+	requiredParams := []string{"name"}
+	err := serviceHelper.RequiredParamsInspector(input, requiredParams)
+	if err != nil {
+		return NewServiceOutput(UserError, err.Error())
+	}
+
+	name, err := valueObject.NewServiceName(input["name"])
+	if err != nil {
+		return NewServiceOutput(UserError, err.Error())
+	}
+
+	var typePtr *valueObject.ServiceType
+	if input["type"] != nil {
+		svcType, err := valueObject.NewServiceType(input["type"])
+		if err != nil {
+			return NewServiceOutput(UserError, err.Error())
+		}
+		typePtr = &svcType
+	}
+
+	var startCmdPtr *valueObject.UnixCommand
+	if input["startCmd"] != nil {
+		startCmd, err := valueObject.NewUnixCommand(input["startCmd"])
+		if err != nil {
+			return NewServiceOutput(UserError, err.Error())
+		}
+		startCmdPtr = &startCmd
+	}
+
+	var statusPtr *valueObject.ServiceStatus
+	if input["status"] != nil {
+		status, err := valueObject.NewServiceStatus(input["status"].(string))
```

Comment:
Why .(string) if NewServiceStatus is interface?

Reply:
Done.